pkg/clusteragent/autoscaling/workload/local/replica_calculator.go

```diff
@@ -0,0 +1,325 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+// Package local provides local recommendations for autoscaling workloads.
+package local
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	datadoghq "github.com/DataDog/datadog-operator/api/datadoghq/v1alpha1"
+
+	workloadmeta "github.com/DataDog/datadog-agent/comp/core/workloadmeta/def"
+
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/autoscaling/workload/common"
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/autoscaling/workload/loadstore"
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/autoscaling/workload/model"
+	le "github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver/leaderelection/metrics"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type replicaCalculator struct {
+	PodWatcher common.PodWatcher
+}
+
+type utilizationResult struct {
+	averageUtilization      float64
+	podToUtilization        map[string]float64
+	missingPods             []string
+	recommendationTimestamp time.Time
+}
+
+func newReplicaCalculator(podWatcher common.PodWatcher) replicaCalculator {
+	return replicaCalculator{
+		PodWatcher: podWatcher,
+	}
+}
+
+// CalculateHorizontalRecommendations is the entrypoint to calculate the horizontal recommendation for a given DatadogPodAutoscaler
+func (r replicaCalculator) CalculateHorizontalRecommendations(dpai model.PodAutoscalerInternal, lStore loadstore.Store) (*model.ScalingValues, error) {
+	currentTime := time.Now()
```

Comment:
nit: I see it's passed as a parameter later to `recommend` but generally speaking to ease testing you can inject a `clock.Clock`, see usage in `controller.go`